### PR TITLE
✨ Added Z3 toggle to the `pyfiction` build process

### PR DESCRIPTION
**Pull request by simon1hofmann**
## Description
Added a toggle to enable Z3 during the `setup.py` build process.

Can be invoked by inserting an environmental variable before the build process:
`z3="ON" pip install .`

Defaults to `z3="OFF"` if not specified. 

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -31,7 +34,7 @@ def build_extension(self, ext):
                       '-DFICTION_TEST=OFF',
                       '-DFICTION_EXPERIMENTS=OFF',
                       '-DFICTION_PYTHON_BINDINGS=ON',
-                      '-DFICTION_Z3=ON',
+                      f'-DFICTION_Z3={z3}',
```
Comment (marcelwa) [inline setup.py:37]: I feel like the default should be OFF to be in line with the fiction flag.

**Review by marcelwa — CHANGES_REQUESTED** [review]
If you could revert the submodule changes, that would be perfect. Many thanks!

**Comment by marcelwa** [comment]
Many thanks for the edits. Can you now adjust the Python workflow accordingly to make sure it builds Z3?

Also is there a solution for the Z3-enabled test cases? Such that they are not run when Z3 is disabled? In our C++ test cases, we have implemented a solution using macros.

File: .github/workflows/python-bindings.yml [hunk 2]
```diff
@@ -68,7 +69,7 @@ jobs:
 
       - name: Build pyfiction
         working-directory: ${{github.workspace}}
-        run: pip install .
+        run: z3=${{env.ENABLE_Z3}} pip install .
```
Comment (marcelwa) [inline .github/workflows/python-bindings.yml:72]: This looks good. Many thanks! 

Regarding the Python tests: right now, they will fail if Z3 was not enabled. Can we disable the respective ones in that case? I was thinking exposing a certain constant in the `exact` bindings and then checking for it in the Python tests for `exact`. If you know of a better way, please let me know 🙂

**Review by marcelwa — CHANGES_REQUESTED** [review]

File: .github/workflows/python-bindings.yml (hunk 2 above)
Comment (simon1hofmann) [reply .github/workflows/python-bindings.yml:72]: My idea would be to save the environment variable `z3` in an `.env` file and skip the respective tests based on that value ?

Comment (marcelwa) [reply .github/workflows/python-bindings.yml:72]: If that's possible, that's even better. Thank you!

File: setup.py [hunk 3]
```diff
@@ -23,6 +23,13 @@ def build_extension(self, ext):
         if not extdir.endswith(os.path.sep):
             extdir += os.path.sep
 
+        # check if Z3 should be used. If no argument is given, fiction depends on Z3 by default.
```
Comment (marcelwa) [inline setup.py:26]: This docstring seems out of date.

**Review by marcelwa — CHANGES_REQUESTED** [review]
Looks really good! Just a tiny inconsistency: one docstring seems to have run out of date.

**Review by marcelwa — APPROVED** [review]
Many thanks for your contribution! LGTM.

**Pull request merged**